src/Fixer/Whitespace/BlankLineBeforeStatementFixer.php

```diff
@@ -328,4 +311,50 @@ protected function createConfigurationDefinition()
                 ->getOption(),
         ]);
     }
+
+    /**
+     * @param int $prevNonWhitespace
+     *
+     * @return bool
+     */
+    private function shouldAddBlankLine(Tokens $tokens, $prevNonWhitespace)
+    {
+        $prevNonWhitespaceToken = $tokens[$prevNonWhitespace];
+
+        if ($prevNonWhitespaceToken->isComment()) {
+            for ($j = $prevNonWhitespace - 1; $j >= 0; --$j) {
+                if (false !== strpos($tokens[$j]->getContent(), "\n")) {
+                    return false;
+                }
+
+                if (!$tokens[$j]->isWhitespace()) {
+                    return !$tokens[$j]->equals('{');
+                }
+            }
+        }
+
+        return $prevNonWhitespaceToken->equalsAny([';', '}']);
+    }
+
+    /**
+     * @param int $index
+     */
+    private function insertBlankLine(Tokens $tokens, $index)
+    {
+        $prevIndex = $index - 1;
+        $prevToken = $tokens[$prevIndex];
+        $lineEnding = $this->whitespacesConfig->getLineEnding();
+
+        if ($prevToken->isWhitespace()) {
+            $countParts = substr_count($prevToken->getContent(), "\n");
```

Comment:
Why parts? I'd expect if the string has one `\n` then it has 2 parts. Can it be `$countNewlines`?

Reply:
TBH I moved this code for readability (and ease the usage of the variable `$index` in the `applyFix` method) and haven't really touch nor studied it